Storybook: Update to latest 5.3

ItsJonQ opened the pull request:
<img width="812" alt="Screen Shot 2020-01-13 at 11 09 22 AM" src="https://user-images.githubusercontent.com/2322354/72272440-9d916580-35f6-11ea-9fbe-80b0e559334e.png">

This update bumps Storybook to the latest 5.3.2 release. Some of the core `.js`
files have been renamed and refactored according to their [5.2.x -> 5.3
migration guides](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#from-version-52x-to-53x).

This update also start some Storybook theming, adding the WordPress branding
logo on the top left.

### #branding

The WP logo is being served from a URL. Originally, I tried to serve the image locally. However, I don't think it's working correctly with Storybook.

https://github.com/storybookjs/storybook/issues/5878
https://github.com/storybookjs/storybook/issues/6521

Maybe able to resolve this with Webpack work-arounds?



## How has this been tested?
Run storybook locally via `npm run storybook:dev`. Also did a local build via `npm run storybook:build`. I tested the local `dist` build of storybook separately to ensure that it will work when built/deployed.


## Types of changes
* Bump Storybook dependencies from `5.2.4` -> `5.3.2`
* Adjusted main Storybook files according to their migration guides
* Added WP logo 🙌 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

Commit:
Storybook: Update to latest 5.3

This update bumps Storybook to the latest 5.3.2 release. Some of the core `.js`
files have been renamed and refactored according to their 5.2.x -> 5.3
migration guides.

This update also start some Storybook theming, adding the WordPress branding
logo on the top left.

epiqueras commented on storybook/main.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+	addons: [
+		'@storybook/addon-docs/react/preset',
```
They use `name: '@storybook/addon-docs',` in the migration guide.

epiqueras replied:
>Addon-docs configuration gets simpler in 5.3. In 5.2, each framework had its own preset, e.g. @storybook/addon-docs/react/preset. Starting in 5.3, everybody should use @storybook/addon-docs/preset.

ItsJonQ replied:
@epiqueras Thank you! 

I'll make those updates. I'm also not 100% clear if the add-ons need to be Objects to work correctly. It seemed to work as strings (at least for me). I imagined it worked similarly to how Babel's config worked.

epiqueras replied:
They can be strings if you don't need extra configuration.

epiqueras commented on storybook/main.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+	addons: [
```
Shouldn't the stories RegExp be here now?

epiqueras commented on storybook/preview.js:
```diff
@@ -19,6 +19,7 @@ import './style.scss';
 addDecorator( withA11y );
 addDecorator( withKnobs );
+
 configure(
```
This should be part of `main.js` now, right?

ItsJonQ replied:
That's my guess as well. However, I wasn't sure how to do it given the conditional `require` we currently have. I'll give it a shot!

epiqueras reviewed (changes requested):
We still need to deal with:

>- We've deprecated the ability to specify the hierarchy separators (how you control the grouping of story kinds in the sidebar). From Storybook 6.0 we will have a single separator /, which cannot be configured.
>- In 5.3, the CSF loader decouples the story's name/id, which means that displayName is no longer necessary. Unfortunately, this is a breaking change for any code that uses the story name field. Storyshots relies on story name, and the appropriate migration is to simply update your snapshots. Apologies for the inconvenience!

Commit:
Update require to use stories config in main.js

Commit:
Update separator to use new Storybook standard

ItsJonQ commented:
@epiqueras I just pushed the updates to the `main.js` config. I've also updated the separators (previously `|`) to the new Storybook standardized `/`.

epiqueras commented:
Thanks, we still need to update the snapshots.

ItsJonQ commented:
@epiqueras Oh! Thanks for the reminder <3
I'm working on that now. Jest is throwing some DOM related errors. Not sure what's causing that. 

```
  ● Storyshots › Components/ClipboardButton › Default

    TypeError: First argument must be a String, HTMLElement, HTMLCollection, or NodeList

      28 | 		const container = this.containerRef.current;
      29 |
    > 30 | 		this.clipboard = new Clipboard( container.firstChild,	{
         | 		                 ^
      31 | 			text: getText,
      32 | 			container,
      33 | 		} );
```

Taking a look at the test file, it looks like there may be something to help with that:

```
			if ( story.kind === 'Components|ClipboardButton' ) {
				currentElement.appendChild(
					document.createElement( 'button' )
				);
			}
``

Will keep investigating

epiqueras commented:
You need to mock the refs.

Commit:
Update Storyshots for testing

ItsJonQ commented:
Rebasing with latest `master`

Commit:
Rebase with latest master

gziolo commented:
https://travis-ci.com/WordPress/gutenberg/jobs/275074009#L398

It looks like types checking errors for some unknown reasons.

ItsJonQ commented:
@gziolo Interesting 🤔. I'm not sure what's up with that. I'll poke at it to see if I can debug.

Thank you!!

Commit:
Fix Jest type issues

This update fixes the Jest type checking issue by installing the latest
@types/jest definitions as a devDependency.

ItsJonQ commented:
@gziolo Halloo!! I think I fixed it 🙏 . I installed the latest `@types/jest` as a devDependency. It looks like the project didn't have it. I think the type definitions were failing because an older version of `@types/jest` may have been a dependency of another dependency.

I based this fix off of this GH issue:
https://github.com/testing-library/jest-dom/issues/152

ItsJonQ commented:
Thanks for the review @epiqueras !! Will merge 💪
